File: project-euler/258/euler_258_v1.cpp
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h>

typedef long long ll;
typedef uint32_t u32;
typedef unsigned char byte;

const u32 BASE = 20092010;
const int LEN = 2000;
const int PL = LEN - 1;

struct mat_look_t
{
    u32 c[LEN][LEN];
};

struct Mat
{
    // original and transposed.
    mat_look_t orig, trans;
};

Mat ONE;

static void mul(Mat * ret, const Mat * const m_base, const Mat * const n_base)
{
    const typeof(&(m_base->orig)) m = &(m_base->orig);
    const typeof(&(n_base->trans)) n = &(n_base->trans);

    for (int r = 0 ; r < LEN ; r++)
    {
        const u32 * mi = (m->c[r]);
        for (int c = 0; c < LEN ; c++)
        {
            ll sum = 0;
            const u32 * ni = (n->c[c]);
            for (int i = 0; i < LEN ; i++)
            {
                sum = ((sum + (((ll)mi[i]) * ((ll)ni[i]))) % BASE);
            }
            ret->trans.c[c][r] = ret->orig.c[r][c] = (u32)sum;
        }
    }
}

static void power_helper(ll n, Mat * ret);

static void power(ll n, Mat * ret)
{
    char cache_fn[200];
    snprintf(cache_fn, sizeof(cache_fn), "CACHE/%llX.bin", n);
    FILE * cache_fh = fopen(cache_fn, "rb");
    if (! cache_fh)
    {
        power_helper(n,ret);
        FILE * w = fopen(cache_fn, "wb");
        fwrite(ret, sizeof(*ret), 1, w);
        fclose(w);
        cache_fh = fopen(cache_fn, "rb");
    }
    if (fread(ret, sizeof(*ret), 1, cache_fh) != 1)
    {
        fprintf(stderr, "SNAFU for %lld\n" , n);
        exit(-1);
    }
    fclose(cache_fh);
}

static void power_helper(ll n, Mat * ret)
{
    if (n == 1)
    {
        *ret = ONE;
    }
    else if (n & 0x1)
    {
        Mat * temp_ret = new Mat;
        power ((n ^ 0x1), temp_ret);
        printf ("After rec %lld\n", n);
        fflush(stdout);
        mul(ret, temp_ret, &ONE);
        printf ("After mul %lld\n", n);
        fflush(stdout);
        delete temp_ret;
    }
    else
    {
        Mat * rec = new Mat;
        power(n >> 1, rec);
        printf ("After rec %lld\n", n);
        fflush(stdout);
        mul(ret, rec, rec);
        printf ("After mul %lld\n", n);
        fflush(stdout);
        delete rec;
    }
}

int main()
{
    for (int r = 0 ; r < LEN ; r++)
    {
        for (int c = 0; c < LEN ; c++)
        {
            ONE.trans.c[c][r] = ONE.orig.c[r][c] =
                (
                    (
                            ((r == PL) && (c <= 1))
                            ||
                            ((r < PL) && (r + 1 == c))
                    ) ? 1 : 0
                );
        }
    }

    Mat * p = new Mat;
#if 1
    power( 1000000000000000000LL, p);
#elif 0
    power( 2LL, p);
#else
    power( 4000LL, p);
#endif

    ll sum = 0;

    for (int c = 0; c < LEN ; c++)
    {
        sum = (sum + p->orig.c[0][c]) % BASE;
    }

    printf ("Result = %lld\n", sum);

    delete p;
    return 0;
}

File: project-euler/258/euler-258-v1.pl
#!/usr/bin/perl

use strict;
use warnings;

use integer;
use bytes;

# use Math::BigInt lib => 'GMP', ':constant';

use List::Util qw(sum);
use List::MoreUtils qw();

STDOUT->autoflush(1);

my $BASE = 20092010;
my $LEN  = 2000;

# Penultimate len.
my $PL = $LEN - 1;

package Mat;

my $ORIG  = 0;
my $TRANS = 1;

sub new
{
    my $o = \( my $foo = '' );
    my $t = \( my $bar = '' );

    my ( $class, $cb ) = @_;

    # The cell index
    my $ci = 0;
    for my $row ( 0 .. $PL )
    {
        my $ti = $row;
        for my $col ( 0 .. $PL )
        {
            vec( $$o, $ci++, 32 ) = vec( $$t, $ti, 32 ) = $cb->( $row, $col );
        }
        continue
        {
            $ti += $LEN;
        }
    }
    return bless [ $o, $t ], $class;
}

sub mul
{
    my ( $m_base, $n_base ) = @_;

    my $m = $m_base->[$ORIG];
    my $n = $n_base->[$TRANS];

    my $mi = -$LEN;
    my $ni;
    return Mat->new(
        sub {
            my ( $r, $c ) = @_;

            if ( $c == 0 )
            {
                $mi += $LEN;
                $ni = 0;
            }
            my $sum = 0;
            for my $i ( 0 .. $PL )
            {
                ( $sum += vec( $$m, $mi + $i, 32 ) * vec( $$n, $ni++, 32 ) ) %=
                    $BASE;
            }

            return $sum;
        }
    );
}

package main;

my $ONE = Mat->new(
    sub {
        my ( $r, $c ) = @_;
        return (
            (
                       ( ( $r == $PL ) and ( $c <= 1 ) )
                    or ( ( $r < $PL )  and ( $r + 1 == $c ) )
            ) ? 1 : 0
        );
    },
);

sub power
{
    my ($n) = @_;

    if ( $n == 1 )
    {
        return $ONE;
    }
    else
    {
        my $rec = power( $n >> 1 );
        print "After rec $n\n";
        my $ret = $rec->mul($rec);
        print "After mul $n\n";
        return ( ( $n & 0x1 ) ? $ret->mul($ONE) : $ret );
    }
}

my $p = power(1_000_000_000_000_000_000);

my $sum = 0;

my $o = $p->[$ORIG];
for my $col ( 0 .. $PL )
{
    ( $sum += vec( $$o, $col, 32 ) ) %= $BASE;
}

print "Result = $sum\n";

=head1 COPYRIGHT & LICENSE

Copyright 2017 by Shlomi Fish

This program is distributed under the MIT / Expat License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut
